File: 5. Longest Palindromic Substring.java
class Solution {
    public String expand(String s, int left, int right){
        int index = -1;
        if(right - left > 1) index = left + 1;
        StringBuilder sb = new StringBuilder();

        while(left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)){
            sb.append(s.charAt(right));
            left--;
            right++;
        }

        StringBuilder ans = new StringBuilder(sb).reverse();
        if(index != -1) ans.append(s.charAt(index));
        ans.append(sb);

        return ans.toString();
    }
    public String longestPalindrome(String s) {
        if(s.length() == 1) return s;
        String ans = "";

        for(int i = 0; i < s.length() - 1; i++){
            String evenPal = "";
            if(s.charAt(i) == s.charAt(i + 1)){
                evenPal = expand(s, i, i + 1);
            }

            String oddPal = expand(s, i - 1, i + 1);

            if(oddPal.length() > evenPal.length() && oddPal.length() > ans.length()) ans = oddPal;
            else if(evenPal.length() > ans.length()) ans = evenPal; 
        }

        return ans;
    }
}

File: 5. Longest Palindromic Substring.cpp
class Solution {
public:
    vector<int> expand(const string& s, int left, int right) {
        while (left >= 0 && right < (int)s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        return {left + 1, right - 1};
    }

    string longestPalindrome(const string& s) {
        int len = 0, ansLeft = 0, ansRight = 0;
        for (int i = 1; i < (int)s.length(); i++) {
            vector<int> boundary1 = expand(s, i - 1, i + 1);
            vector<int> boundary2(2, 0);
            if (s[i] == s[i - 1]) boundary2 = expand(s, i - 2, i + 1);
            int len1 = boundary1[1] - boundary1[0];
            int len2 = boundary2[1] - boundary2[0];
            if (len < len1 || len < len2) {
                if (len1 < len2) {
                    len = len2;
                    ansLeft = boundary2[0];
                    ansRight = boundary2[1];
                } else {
                    len = len1;
                    ansLeft = boundary1[0];
                    ansRight = boundary1[1];
                }
            }
        }

        return s.substr(ansLeft, ansRight - ansLeft + 1);
    }
};

